pkg/executioncluster/impl/random_cluster_selector.go

```diff
@@ -123,51 +107,72 @@ func (s RandomClusterSelector) GetAllValidTargets() []executioncluster.Execution
 	return v
 }
 
-func (s RandomClusterSelector) GetTarget(spec *executioncluster.ExecutionTargetSpec) (*executioncluster.ExecutionTarget, error) {
-	if spec == nil || (spec.TargetID == "" && spec.ExecutionID == nil) {
-		return nil, fmt.Errorf("invalid executionTargetSpec %v", spec)
+func (s RandomClusterSelector) GetTarget(ctx context.Context, spec *executioncluster.ExecutionTargetSpec) (*executioncluster.ExecutionTarget, error) {
+	if spec == nil {
+		return nil, fmt.Errorf("empty executionTargetSpec")
 	}
 	if spec.TargetID != "" {
 		if val, ok := s.executionTargetMap[spec.TargetID]; ok {
 			return &val, nil
 		}
 		return nil, fmt.Errorf("invalid cluster target %s", spec.TargetID)
 	}
-	if spec.ExecutionID != nil {
-		if weightedRandomList, ok := s.domainWeightedRandomMap[spec.ExecutionID.GetDomain()]; ok {
-			executionName := spec.ExecutionID.GetName()
-			if executionName != "" {
-				randSrc, err := getRandSource(executionName)
-				if err != nil {
-					return nil, err
-				}
-				result, err := weightedRandomList.GetWithSeed(randSrc)
-				if err != nil {
-					return nil, err
-				}
-				execTarget := result.(executioncluster.ExecutionTarget)
-				return &execTarget, nil
-			}
-			execTarget := weightedRandomList.Get().(executioncluster.ExecutionTarget)
-			return &execTarget, nil
+	resource, err := s.resourceManager.GetResource(ctx, managerInterfaces.ResourceRequest{
+		Project:      spec.Project,
+		Domain:       spec.Domain,
+		Workflow:     spec.Workflow,
+		LaunchPlan:   spec.LaunchPlan,
+		ResourceType: admin.MatchableResource_EXECUTION_CLUSTER_LABEL,
+	})
+	if err != nil {
+		if flyteAdminError, ok := err.(errors.FlyteAdminError); !ok || flyteAdminError.Code() != codes.NotFound {
+			return nil, err
 		}
+		return nil, err
 	}
-	return nil, fmt.Errorf("invalid executionTargetSpec %v", *spec)
+	var weightedRandomList random.WeightedRandomList
+	if resource != nil && resource.Attributes.GetExecutionClusterLabel() != nil {
+		label := resource.Attributes.GetExecutionClusterLabel().Value
 
+		if _, ok := s.labelWeightedRandomMap[label]; ok {
+			weightedRandomList = s.labelWeightedRandomMap[label]
+		}
+	}
+	// If there is no label associated (or) if the label is invalid, choose from all enabled clusters.
```

Comment:
should we log when the label is invalid?

Reply:
Sure